### DISCUSSION TITLE
What do you think of splitting CSS into multiple files?

### DISCUSSION THREAD

**Discussion by servitecbaires** [datenstrom/community#777] [Ask a question] [answered]
Hello everyone!!!
My name is Martin from Argentina.
I'll do my best with my English skills + Google translator.

I'm not sure if this is 100% a bug, but after trying several approaches, finally decided to report it.
Well... this is the problem:
I'm making a website based on the new Wittstock theme by @schulle4u
Then I wrote additional css in a separated file (just for convenience) placed in the same theme folder. In fact is the wittstock-custom.css file provided by the theme itself.

In first place, I thinked that theme's name prefixed files will automagically loaded inside header and buldled. That didn't appened.
Later i found this datenstrom/community#699 . Once again @schulle4u gave me (gave to all of us) a clear and straightfoward answer. But it aso didn't work for me.
And... later (many hours later, indeed) I finally wrote a few lines of code to retrieve via API the correct file path to the css file. it didn't work either.

None of these approaches generates a new version of the bundle css (checking the bundle***.css file name. But, if I change a single word on the wittstock.css file, the bundle file name changes, as well.

Long story, short story... this is the code included in the html header:
```
<link rel="stylesheet" type="text/css" href="<?php echo $this->yellow->system->get("coreServerBase").$this->yellow->system->get("coreThemeLocation")."stockholm.css"; ?>" />
<link rel="stylesheet" type="text/css" media="all" href="/media/themes/wittstock-custom.css" />
<link rel="stylesheet" type="text/css" media="all" href="/system/themes/wittstock-custom.css" />
<link rel="stylesheet" type="text/css" href="<?php echo $this->yellow->system->get("coreServerBase").$this->yellow->system->get("coreThemeLocation")."wittstock-custom.css"; ?>" />
<?php 
$path = $this->yellow->system->get("coreThemeDirectory");
$themeName = $this->yellow->system->get("theme");
$regex = "/" . $themeName . "-" . ".*" . "\\.css/";
foreach ($this->yellow->toolbox->getDirectoryEntries($path, $regex, true, false) as $entry) {
     ?><link rel="stylesheet" type="text/css" media="all" href="<?php echo $entry; ?>" /><?php } ?>
<?php echo $this->yellow->page->getExtra("header") ?>
```

I'm using the last version of Yellow and Wittstock theme (downloaded yesterday) 
Running on Kubuntu 20.04 
Selfserved by Datenstrom Yellow Serve extension itself.

Many thanks to everyone for your time and dedication, not only in developing this amazing piece of software, also for the exhaustive support.

Regards!

**Comment by schulle4u** [comment]
Hello Martin, 

thanks for your report. I have published an update for the [Wittstock extension](https://github.com/schulle4u/yellow-wittstock) which now uses Yellow's API to correctly include additional CSS content. The file still is located in your `system/themes` directory. Previously the `wittstock-custom.css` was included using a `@import` variable in the main css file. Please let me know if this works better for you now. 

Steffen

**Reply by servitecbaires** [reply]
Hello Steffen!!

Thanks for your fast reply!
I'm aware that you can bundle the additional css via the `@import `method within the main css file.
Maybe I misunderstood how the Bundle extension works. My goal (or the meaning of this bug report) is to get all the css files included in the page's (by any method) bundled by the Bundle extension.
But... why this way instead the `@import` method? (you might ask)
Because I want to have the chance to modify the styles by a separate css file (prefixed with the theme name), leaving the original without any modification.

Martin

**Reply by schulle4u** [reply]
I'm afraid its a little difficult and confusing to track down the issue. Can you confirm the behaviour in a clean installation using the default stockholm theme? If yes, please give a step-by-step explanation allowing us to understand what you're trying to achieve.

**Comment by servitecbaires** [comment]
Excuse me for beeing so stupid!!!
I didn't notice that Wittstock Theme came with a `wittstock-header.html` file. I was modifying the default `header.html` instead.
That's why the Bundle extension never get any additional css file to bundle.
I'm so ashamed!!
I will keep this thread for 48hs. only with the purpose that all of you can read my apologies for wasting your time.
I will code and share an extension which will load any css (and js file, eventually) prefied with the theme name in the theme directory, automatically. You'll have to be patient, because I have zero skills and knowledge about php. 
I also will share the code to achieve the same goal including it in the theme's `header.html` file.

Regards,
Martin

**Reply by schulle4u** [reply]
It's OK to not know everything. 🙂 Indeed every theme can have it's own layout files. Wittstock uses custom layouts for more semantic HTML code. 

The offer is still valid. A step-by-step explanation helps us to understand your goals. Some solutions turn out to be much simpler than expected, and maybe there's even a bug with the bundle extension we are not aware of? In any case you'll get enhanced knowledge. 🙂

**Comment by servitecbaires** [comment, marked as answer]
OK! This is the step-by-step explanation. (well, to be honest, I'm not really sure about thw last sentence)
Imagine you are using a theme like Stockholm. But also you want/like/need to make some changes to the css.
The goal is to not modify the original css file (neither the theme's header layout) using a theme's name prefixed additional css file, like `stockholm-custom.css` or `stockholm-whatever.css`. Why? (Anyone might ask) Because is simpler to find your additions (delete or them, etc) , instead of diving in the original css file. Is also an easy way to face a re installation, when you only have to copy and paste a css file. This is in a beginner's point of view.
That was my starting point.

1. I assumed that Yellow will auto include any theme's name prefixed css file, like with layouts. I was wrong.
2. I found (in the Discussion forum) that `ThemeLocation` (/system/themes/) doesn't return the url/path which Yellow use to upload the ccs (/media/themes/)file. 
3. Now, I have a code that will search and upload any css file located in `ThemeDirectory` which is prefixed with the theme's name. This code have to be placed after the `getExtra->("header")` function call, because the Bundle Extension will bundle the css (and js) files in order of appearance, so you need to get the theme's default css file first.

This is my code:
```
<?php 
$path = $this->yellow->system->get("coreServerBase").$this->yellow->system->get("CoreThemeDirectory");
$themeName = $this->yellow->system->get("theme");
$regex = "/" . $themeName . "-" . ".*" . "\\.css/";
foreach ($this->yellow->toolbox->getDirectoryEntries($path, $regex, false, false, false) as $entry) { ?>
<link rel="stylesheet" type="text/css" media="all" href="<?php echo this->yellow->system->get("coreServerBase").$this->yellow->system->get("coreThemeLocation").$entry; ?>" /><?php } ?>
```
I will find a way to make an extension which can include this code when you call the `getExtra("header")` function.
Or maybe (this is me wishing things aloud) this could be included as a feature in Yellow, allowing include additional css without editing any layout file or css file.

Thanks again for the time that all of you dedicate in the dev and support of this amazing program. Also for the time you dedicate reading this lines.
Once again, I feel this thread is off-topic here in Bug's section. I will leave it 48hs for you to read it, and then I will delete it and re-share the code again in the proper section, in case that could be useful for others.

Regards.
Martín

**Reply by schulle4u** [reply]
No need to worry. We blandly ask people to edit their topics or move them in place if something's wrong. 🙂

OK, I think I've understand your problem now. Why not make a code proposal or a pull request for the actual `getExtra()`function in the core extension? This would take out unnecessary complexity for new users. The problem however could be that different features/themes use different methods to include additional files. An automated recognition therefore could have unwanted side effects. I personally prefer as few files as possible, including everything else only when needed for individual extensions.
